clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java

```diff
@@ -1068,6 +1072,149 @@ public void testUnsupportedCompress() {
+        MemoryRecords records = createMemoryRecords(args, recordsPerOffset);
+
+        // Test slicing from start
+        MemoryRecords sliced = records.slice(0, records.sizeInBytes());
+        assertEquals(records.sizeInBytes(), sliced.sizeInBytes());
+        assertEquals(records.validBytes(), sliced.validBytes());
+        TestUtils.checkEquals(records.batches(), sliced.batches());
+
+        List<RecordBatch> items = batches(records);
+        // Test slicing first message.
+        RecordBatch first = items.get(0);
+        sliced = records.slice(first.sizeInBytes(), records.sizeInBytes() - first.sizeInBytes());
+        assertEquals(records.sizeInBytes() - first.sizeInBytes(), sliced.sizeInBytes());
+        assertEquals(items.subList(1, items.size()), batches(sliced), "Read starting from the second message");
+        assertTrue(sliced.validBytes() <= sliced.sizeInBytes());
+
+        // Read from second message and size is past the end of the file.
+        sliced = records.slice(first.sizeInBytes(), records.sizeInBytes());
+        assertEquals(records.sizeInBytes() - first.sizeInBytes(), sliced.sizeInBytes());
+        assertEquals(items.subList(1, items.size()), batches(sliced), "Read starting from the second message");
+        assertTrue(sliced.validBytes() <= sliced.sizeInBytes());
+
+        // Read from second message and position + size overflows.
+        sliced = records.slice(first.sizeInBytes(), Integer.MAX_VALUE);
+        assertEquals(records.sizeInBytes() - first.sizeInBytes(), sliced.sizeInBytes());
+        assertEquals(items.subList(1, items.size()), batches(sliced), "Read starting from the second message");
+        assertTrue(sliced.validBytes() <= sliced.sizeInBytes());
+
+        // Read a single message starting from second message.
+        RecordBatch second = items.get(1);
+        sliced = records.slice(first.sizeInBytes(), second.sizeInBytes());
+        assertEquals(second.sizeInBytes(), sliced.sizeInBytes());
+        assertEquals(Collections.singletonList(second), batches(sliced), "Read a single message starting from the second message");
+
+        // Read from already sliced view.
+        List<RecordBatch> remainingItems = IntStream.range(0, items.size()).filter(i -> i != 0 && i != 1).mapToObj(items::get).collect(Collectors.toList());
+        int remainingSize = remainingItems.stream().mapToInt(RecordBatch::sizeInBytes).sum();
+        sliced = records.slice(first.sizeInBytes(), records.sizeInBytes() - first.sizeInBytes())
+                        .slice(second.sizeInBytes(), records.sizeInBytes() - first.sizeInBytes() - second.sizeInBytes());
+        assertEquals(remainingSize, sliced.sizeInBytes());
+        assertEquals(remainingItems, batches(sliced), "Read starting from the third message");
+
+        // Read from second message and size is past the end of the file on the already sliced view.
+        sliced = records.slice(1, records.sizeInBytes() - 1)
```

Comment:
The expected usage is that position will be aligned with batch boundary, right?

Reply:
Yes, that's true. And test case also verifies that though the first slice was random but subequent slice on same with correct batch boundary resulted in correct data.